core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c

```diff
@@ -994,51 +1181,38 @@ wasi_sock_recv(wasm_exec_env_t exec_env, wasi_fd_t sock, iovec_app_t *ri_data,
     wasm_module_inst_t module_inst = get_module_inst(exec_env);
     wasi_ctx_t wasi_ctx = get_wasi_ctx(module_inst);
     struct fd_table *curfds = wasi_ctx_get_curfds(module_inst, wasi_ctx);
-    wasi_iovec_t *iovec, *iovec_begin;
     uint64 total_size;
-    size_t ro_datalen;
     uint32 i;
     wasi_errno_t err;
 
     if (!wasi_ctx)
-        return (wasi_errno_t)-1;
+        return __WASI_EINVAL;
 
     total_size = sizeof(iovec_app_t) * (uint64)ri_data_len;
     if (!validate_native_addr(ro_datalen_app, (uint32)sizeof(uint32))
         || !validate_native_addr(ro_flags, (uint32)sizeof(wasi_roflags_t))
         || total_size >= UINT32_MAX
         || !validate_native_addr(ri_data, (uint32)total_size))
-        return (wasi_errno_t)-1;
-
-    total_size = sizeof(wasi_iovec_t) * (uint64)ri_data_len;
-    if (total_size >= UINT32_MAX
-        || !(iovec_begin = wasm_runtime_malloc((uint32)total_size)))
-        return (wasi_errno_t)-1;
+        return __WASI_EINVAL;
 
-    iovec = iovec_begin;
+    /* recv ri_data one by one */
+    for (i = 0; i < ri_data_len; ri_data++, i++) {
+        void *buf;
 
-    for (i = 0; i < ri_data_len; i++, ri_data++, iovec++) {
         if (!validate_app_addr(ri_data->buf_offset, ri_data->buf_len)) {
-            err = (wasi_errno_t)-1;
-            goto fail;
+            return __WASI_EINVAL;
         }
-        iovec->buf = (void *)addr_app_to_native(ri_data->buf_offset);
-        iovec->buf_len = ri_data->buf_len;
-    }
-
-    err = wasmtime_ssp_sock_recv(curfds, sock, iovec_begin, ri_data_len,
-                                 ri_flags, &ro_datalen, ro_flags);
-    if (err)
-        goto fail;
-
-    *(uint32 *)ro_datalen_app = (uint32)ro_datalen;
 
-    /* success */
-    err = 0;
+        buf = (void *)addr_app_to_native(ri_data->buf_offset);
+        err = wasmtime_ssp_sock_recv(curfds, sock, buf, ri_data->buf_len);
+        if (err != __WASI_ESUCCESS) {
+            return err;
+        }
+    }
 
-fail:
-    wasm_runtime_free(iovec_begin);
-    return err;
+    *(uint32 *)ro_datalen_app = ri_data_len;
```

Comment:
ri_data_len seems to be the packet count, is it right to set it (but not total ri_data->buf_len) to ro_datalen_app?

Reply:
Right. should make it right.

Reply:
OK